quinn/src/incoming.rs

```diff
@@ -0,0 +1,149 @@
+use std::{
+    fmt,
+    future::{Future, IntoFuture},
+    net::{IpAddr, SocketAddr},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use bytes::BytesMut;
+use proto::ConnectionError;
+use thiserror::Error;
+
+use crate::{
+    connection::{Connecting, Connection},
+    endpoint::EndpointRef,
+};
+
+/// An incoming connection for which the server has not yet begun its part of the handshake
+pub struct Incoming(Option<State>);
+
+impl Incoming {
+    pub(crate) fn new(
+        inner: proto::Incoming,
+        endpoint: EndpointRef,
+        response_buffer: BytesMut,
+    ) -> Self {
+        Self(Some(State {
+            inner,
+            endpoint,
+            response_buffer,
+        }))
+    }
+
+    /// Attempt to accept this incoming connection (an error may still occur)
+    pub fn accept(mut self) -> Result<Connecting, ConnectionError> {
+        let state = self.0.take().unwrap();
+        state.endpoint.accept(state.inner, state.response_buffer)
+    }
+
+    /// Reject this incoming connection attempt
+    pub fn reject(mut self) {
```

Comment:
If the application simply drops the Incoming and not calling reject, will the server still bother sending the INITIAL_CLOSE?

Reply:
The `Drop` impl just below calls `reject` automatically. Call `ignore` if you don't want that.